src/ParticleContainer.cpp

```diff
@@ -54,12 +42,365 @@ std::vector<Particle>::const_iterator ParticleContainer::end() const {
+                 r_cutoff);
+  }
+}
+
+void LCParticleContainer::handleLJFCalculation() {
+  std::array<int, 6> boundarys = {1,1,1,1,1,1};
+  realocateParticles();
+  int empty_counter = 0;
+  for (auto &c : cells) {
+    if (!c.isEmpty()) {
+      std::vector<Particle> neighbourhood =
+          getParticleInNeighbourhood(c.getId());
+      LCParticleContainer container;
+      Calculations calc(container);
+      calc.LCcalculateLJF(c.getParticles(), neighbourhood);
+
+    } else {
+      empty_counter++;
+    }
+  }
+  spdlog::debug("there where {} empty cells", empty_counter);
+}
+
+bool LCParticleContainer::addParticleToCell(Particle &p) {
+  int x = int(floor(p.getX().at(0) / cell_size.at(0)));
+  int y = int(floor(p.getX().at(1) / cell_size.at(1)));
+  int z = int(floor(p.getX().at(2) / cell_size.at(2)));
+  if (cellExists({x, y, z})) {
+    getCellById({x, y, z}).addParticle(&p);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::vector<Particle> &LCParticleContainer::getParticles() { return particles; }
+
+void LCParticleContainer::addMultipleParticles(std::vector<Particle> &newParticles) {
+  for (auto &p : newParticles) {
+    particles.push_back(p);
+  }
+}
+
+void LCParticleContainer::addParticle(Particle p) {
+  particles.push_back(p);
+  addParticleToCell(p);
+}
+
+std::vector<Cell> LCParticleContainer::getCells() { return cells; }
+
+bool LCParticleContainer::cellExists(std::array<int, 3> id) {
+  for (auto c : cells) {
+    if (c.getId().at(0) == id.at(0) && c.getId().at(1) == id.at(1) &&
+        c.getId().at(2) == id.at(2)) {
+      return true;
+    }
+  }
+  return false;
+};
+
+void LCParticleContainer::countParticlesInCells() {
+  int counter = 0;
+  for(auto c: cells) {
+    for(auto p: c.getParticles()) {
+      counter ++;
+    }
+  }
+  spdlog::info("there are {} particles in all the cells", counter);
+} //just for debugging
+std::vector<Particle *> LCParticleContainer::getBoundaryParticles() {
+  std::vector<Particle*> result;
+  for(int x = 0; x < cell_count[0]; x++) {
+    for(int y = 0; y < cell_count[1]; y++) {
+      //for(int z = 0; z < cell_count[2]; z++) {
+        if(x==0 || x==(cell_count[0]-1) || y==0 || y==(cell_count[1]-1) ) { // || z==0 || z==(cell_count[2]-1)
+          for(auto p : getCellById({x,y,0}).getParticles()) {
+            result.push_back(p);
+          }
+        }
+      }
+    //}
+  }
+  return result;
+}
+
+void LCParticleContainer::handleBoundaryAction() {
+  std::vector<Particle*> boundaryparticles = getBoundaryParticles();
+  if(boundaryparticles.size()>0) {
+    for(auto p: boundaryparticles) {
+      std::array<double, 6> bounds = getInfluencingBoundarysWithDistance(p);
+      std::array<double, 6> compare = {-1.0, -1.0, -1.0, -1.0, -1.0, -1.0};
+      if(bounds != compare) {
+        for(int i = 0; i < 6; i++) {
+          if(boundary_types[i] == 2){
+            // generate halo Particle
+            if(bounds.at(i) != -1.0 && bounds.at(i) <= (pow(2, 1/6)/2)){
```

Comment:
The complexity of this can surely be reduced.
```suggestion
for(auto p: boundaryparticles) {
  std::array<double, 6> bounds = getInfluencingBoundarysWithDistance(p);
    for(int i = 0; i < 6; i++) {
        if(boundary_types[i] == 2){
          // generate halo Particle
          if(bounds.at(i) != -1.0 && bounds.at(i) <= (pow(2, 1/6)/2)){
```
e.g. the first size check is unnecessary.
Also, I don't think you gain much with the `compare` check